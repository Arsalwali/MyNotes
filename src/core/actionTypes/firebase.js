const prefix = 'firebase';

export const ADD_NOTE = `${prefix}/ADD_NOTE`;
export const EDIT_NOTE = `${prefix}/EDIT_NOTE`;
export const DELETE_NOTE = `${prefix}/DELETE_NOTE`;
