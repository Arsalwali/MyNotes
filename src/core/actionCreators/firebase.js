import { ADD_NOTE, EDIT_NOTE, DELETE_NOTE } from '../actionTypes/firebase';

export const addNote = (payload) => ({
  type: ADD_NOTE,
  payload,
});

export const editNote = (payload) => ({
  type: EDIT_NOTE,
  payload,
});

export const deleteNote = (payload) => ({
  type: DELETE_NOTE,
  payload,
});



