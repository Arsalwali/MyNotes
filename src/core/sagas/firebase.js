import { call, put, takeEvery } from 'redux-saga/effects';
import firebase from 'react-native-firebase';
import { ADD_NOTE, EDIT_NOTE, DELETE_NOTE } from '../actionTypes/firebase';

export function* firebaseSaga() {
  const refs = firebase.firestore().collection('notes');
  yield takeEvery([ADD_NOTE, EDIT_NOTE, DELETE_NOTE], callFirestore, refs);
}

export function* callFirestore(refs, action) {
  const { key, title, note, color } = action.payload;
  if (action.type === ADD_NOTE) {
    refs.add({
      title,
      note,
      color,
    }).then(() => console.warn('add success'))
    .catch((error) => console.warn('add error'));
  } else if (action.type === EDIT_NOTE) {
    refs.doc(key).update({
      title,
      note,
      color,
    }).then(() => console.warn('edit success'))
    .catch((error) => console.warn('edit error'));
  } else if (action.type === DELETE_NOTE) {
    refs.doc(key).delete()
    .then(() => console.warn('delete success'))
    .catch((error) => console.warn('delete error'));
  }
}