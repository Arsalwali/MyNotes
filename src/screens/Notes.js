import React from 'react';
import { Text, View, Button, FlatList } from 'react-native';
import NoteModal from './NoteModal';
import firebase from 'react-native-firebase';
import { connect } from 'react-redux';
import { showModal } from './actionCreators';
import { deleteNote } from '../core/actionCreators/firebase';
import NoteCard from './NoteCard';
import { Header } from '../components';
import { colors, iconNames } from '../theme';

export class Notes extends React.Component {
  constructor(props) {
    super(props);
    this.ref = firebase.firestore().collection('notes');
    this.unsubscribe = null;
  }

  componentDidMount() {
    this.unsubscribe = this.ref.onSnapshot(this.onCollectionUpdate);
  }

  componentWillUnmount() {
    this.unsubscribe();
  }

  state = {
    notes: [],
  }

  onCollectionUpdate = (querySnapshot) => {
    const notes = [];
    querySnapshot.forEach((doc) => {
      const { title, note, color } = doc.data();
  
      notes.push({
        key: doc.id,
        doc, // DocumentSnapshot
        title,
        note,
        color,
      });
    });
  
    this.setState({ notes });
  }

  onAddPress = () => {
    this.props.showModal(null, '', '');
  }

  onEditPress = (key, title, note, color) => () => {
    this.props.showModal(key, title, note, color);
  }

  onDeletePress = (key) => () => {
    this.props.deleteNote({ key });
  }

  keyExtractor = (item) => item.key;

	renderItem = (data) => (
    <NoteCard
      data={data.item}
      onEditPress={this.onEditPress(data.item.key, data.item.title, data.item.note, data.item.color)}
      onDeletePress={this.onDeletePress(data.item.key)}
    />
	);
	
	render() {
		return (
      <View style={{ flex: 1, backgroundColor: colors.backgroundColor }}>
        <NoteModal />
        <Header title="My Notes" rightIcon={iconNames.add} onIconPress={this.onAddPress} />
        <FlatList
          data={this.state.notes}
          renderItem={this.renderItem}
          keyExtractor={this.keyExtractor}
        />
      </View>
		);
	}
}


const mapStateToProps = (state) => {
  return null;
};

const mapDispatchToProps = { showModal, deleteNote };

export default connect(null, mapDispatchToProps)(Notes);