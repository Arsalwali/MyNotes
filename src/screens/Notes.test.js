import React from 'react';
import { shallow } from 'enzyme';
import {Notes} from './Notes';
import firebase from 'react-native-firebase';

const notesArray = [
  {
    key: '123',
    title: 'First',
    note: "Note",
    color: '#606060',  
  },
  {
    key: '1234',
    title: 'Seccond',
    note: "Note",
    color: '#60F060',  
  }
];

const props = {
  showModal: jest.fn(),
  deleteNote: jest.fn(),
};


beforeEach(() => {
 
  // const firestore = firebase.firestore = jest.fn();
});

describe('Notes screen', () => {
  it('should render properly', () => {
    const wrapper = shallow(<Notes {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render properly with notes', () => {
    const wrapper = shallow(<Notes {...props} />);
    wrapper.setState({ notes: notesArray })
    expect(wrapper).toMatchSnapshot();
  });

  it('renderItem() should render one item properly', () => {
    const wrapper = shallow(<Notes {...props} />);
    const instance = wrapper.instance();
    const data ={
      item: {
        key: '123',
        title: 'First',
        note: 'Note',
        color: '#454545',
      }
    }
    const item = instance.renderItem(data);
    expect(item).toMatchSnapshot();
  });

  it('keyExtractor() should render key properly', () => {
    const wrapper = shallow(<Notes {...props} />);
    const instance = wrapper.instance();
    const data ={
      item: {
        key: '123',
        title: 'First',
        note: 'Note',
        color: '#454545',
      }
    }
    const key = instance.keyExtractor(data.item);
    expect(key).toEqual('123');
  });

  it('onAddPress() should show modal with empty text', () => {
    const wrapper = shallow(<Notes {...props} />);
    const instance = wrapper.instance();
    expect(props.showModal.mock.calls.length).toBe(0);
    const key = instance.onAddPress();
    expect(props.showModal.mock.calls.length).toBe(1);
    expect(props.showModal.mock.calls[0][0]).toEqual(null);
    expect(props.showModal.mock.calls[0][1]).toEqual('');
    expect(props.showModal.mock.calls[0][2]).toEqual('');
  });

  it('onEditPress() should show modal with filled text', () => {
    const wrapper = shallow(<Notes {...props} />);
    const instance = wrapper.instance();
    expect(props.showModal.mock.calls.length).toBe(1);
    const key = instance.onEditPress('123', 'Imp', 'long note', '#654542')();
    expect(props.showModal.mock.calls.length).toBe(2);
    expect(props.showModal.mock.calls[1][0]).toEqual('123');
    expect(props.showModal.mock.calls[1][1]).toEqual('Imp');
    expect(props.showModal.mock.calls[1][2]).toEqual('long note');
    expect(props.showModal.mock.calls[1][3]).toEqual('#654542');
    
  });

  it('onDeletePress() should call delete note', () => {
    const wrapper = shallow(<Notes {...props} />);
    const instance = wrapper.instance();
    expect(props.deleteNote.mock.calls.length).toBe(0);
    const key = instance.onDeletePress('123')();
    expect(props.deleteNote.mock.calls.length).toBe(1);
    expect(props.deleteNote.mock.calls[0][0]).toEqual({ key: '123' });
  });

  it('onCollectionUpdate() should set notes', () => {
    const note = {
      key: '123', title: '1234', note: '1234455', color: '#232DD9',
    }
    const firebaseData = {
      querySnapshot: [
        {
            data: jest.fn(() => note),
            id: '123',
        }
      ],
    };
    const wrapper = shallow(<Notes {...props} />);
    const instance = wrapper.instance();
    instance.onCollectionUpdate (firebaseData.querySnapshot);
    expect(wrapper.state().notes).toEqual([note]);
  });
});
