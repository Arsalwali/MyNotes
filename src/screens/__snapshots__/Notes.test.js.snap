// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notes screen renderItem() should render one item properly 1`] = `
<NoteCard
  data={
    Object {
      "color": "#454545",
      "key": "123",
      "note": "Note",
      "title": "First",
    }
  }
  onDeletePress={[Function]}
  onEditPress={[Function]}
/>
`;

exports[`Notes screen should render properly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#dce2e5",
      "flex": 1,
    }
  }
>
  <Connect(NoteModal) />
  <Header
    onIconPress={[Function]}
    rightIcon="add"
    title="My Notes"
  />
  <FlatList
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;

exports[`Notes screen should render properly with notes 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#dce2e5",
      "flex": 1,
    }
  }
>
  <Connect(NoteModal) />
  <Header
    onIconPress={[Function]}
    rightIcon="add"
    title="My Notes"
  />
  <FlatList
    data={
      Array [
        Object {
          "color": "#606060",
          "key": "123",
          "note": "Note",
          "title": "First",
        },
        Object {
          "color": "#60F060",
          "key": "1234",
          "note": "Note",
          "title": "Seccond",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;
